benches/dapol.rs

```diff
@@ -1,175 +1,350 @@
-use core::fmt::Debug;
-use criterion::{criterion_group, criterion_main, BatchSize, BenchmarkId, Criterion, SamplingMode};
-use dapol::{
-    utils::get_secret, Dapol, DapolNode, RangeProofPadding, RangeProofSplitting, RangeProvable,
-    RangeVerifiable,
-};
-use digest::Digest;
-use rand::{distributions::Uniform, thread_rng, Rng};
-use smtree::{
-    index::TreeIndex,
-    traits::{ProofExtractable, Rand, Serializable, TypeName},
-};
-use std::time::Duration;
-
-// CONSTANTS
-// ================================================================================================
+mod setup;
+
+use criterion::{criterion_group, criterion_main};
+use criterion::{BenchmarkId, Criterion};
+use iai_callgrind::{black_box, library_benchmark, library_benchmark_group, main};
+
+// use std::path::PathBuf;
+use std::str::FromStr;
 
-const TREE_HEIGHTS: [usize; 3] = [16, 24, 32];
-const NUM_LEAVES: [usize; 3] = [1024, 2048, 4096];
+use dapol::{EntityId, Height, InclusionProof, MaxThreadCount};
 
-// BENCHMARKS
+use setup::{NUM_USERS, TREE_HEIGHTS};
+
+// BENCHMARKS: CRITERION
 // ================================================================================================
 
-fn build_dapol(c: &mut Criterion) {
+fn bench_build_tree(c: &mut Criterion) {
     let mut group = c.benchmark_group("build");
     group.sample_size(10);
-    group.sampling_mode(SamplingMode::Flat);
-    group.measurement_time(Duration::from_secs(20));
-
-    // bench tree height = 16
-    let tree_height = 16;
-    for &num_leaves in NUM_LEAVES.iter() {
-        let items = build_item_list(num_leaves, tree_height);
-        group.bench_function(BenchmarkId::new("height_16", num_leaves), |bench| {
+
+    let num_entities = NUM_USERS[2]; // 30_000: max. value for tree height 16
+    let thread_counts: [u8; 7] = [4, 8, 16, 32, 64, 128, 255];
```

Comment:
From the issue I wrote: 
> This number is limited by the available parallelism of the machine. We don't want to have too many threads. The highest this number should go is

So this means this `thread_counts` list must be dynamically generated based on the underlying machine.